libdrgn/python/helpers.c

```diff
@@ -249,3 +250,250 @@ PyObject *drgnpy_linux_helper_pgtable_l5_enabled(PyObject *self, PyObject *args,
 		return PyErr_Format(PyExc_ValueError, "not Linux kernel");
 	Py_RETURN_BOOL(prog->prog.vmcoreinfo.pgtable_l5_enabled);
 }
+
+static void GenericIterator_dealloc(GenericIterator *self)
+{
+	if (self->iterator) {
+		self->iterator_deinit(self->iterator);
+		free(self->iterator);
+	}
+	Py_DECREF(self->prog);
+	Py_TYPE(self)->tp_free((PyObject *)self);
+}
+
+static PyObject *GenericIterator_next(GenericIterator *self)
+{
+	return self->next(self);
+}
+
+PyTypeObject GenericIterator_type = {
+	PyVarObject_HEAD_INIT(NULL, 0)
+	.tp_name = "_drgn._GenericIterator",
+	.tp_basicsize = sizeof(GenericIterator),
+	.tp_dealloc = (destructor)GenericIterator_dealloc,
+	.tp_flags = Py_TPFLAGS_DEFAULT,
+	.tp_iter = PyObject_SelfIter,
+	.tp_iternext = (iternextfunc)GenericIterator_next,
+};
+
+static PyObject *for_each_task_next(GenericIterator *self)
+{
+	struct drgn_error *err;
+	struct task_iter *iterator = (struct task_iter *)self->iterator;
+	struct drgn_object *entry;
+	err = task_iter_next(iterator, &entry);
+	if (err)
+		return set_drgn_error(err);
+	if (!entry)
+		return NULL;
+	DrgnObject *ret = DrgnObject_alloc(self->prog);
+	if (!ret)
+		return NULL;
+	err = drgn_object_copy(&ret->obj, entry);
+	if (err)
+		return set_drgn_error(err);
+	return (PyObject *)ret;
+}
+
+GenericIterator *drgnpy_linux_helper_for_each_task(PyObject *self,
+						   PyObject *args,
+						   PyObject *kwds)
+{
+	static char *keywords[] = {"prog_or_ns", NULL};
+	struct drgn_error *err;
+	struct prog_or_ns_arg prog_or_ns;
+
+	if (!PyArg_ParseTupleAndKeywords(args, kwds, "O&:for_each_task",
+					 keywords, &prog_or_pid_ns_converter,
+					 &prog_or_ns))
+		return NULL;
+
+	GenericIterator *iterator =
+		(GenericIterator *)GenericIterator_type.tp_alloc(
+			&GenericIterator_type, 0);
+	if (!iterator)
+		return NULL;
+	iterator->next = for_each_task_next;
+	iterator->iterator_deinit = (void (*)(void *))task_iter_deinit;
+	iterator->iterator = malloc(sizeof(struct task_iter));
+	if (!iterator->iterator) {
+		Py_DECREF(iterator);
+		return set_drgn_error(&drgn_enomem);
+	}
+	err = task_iter_init(iterator->iterator, prog_or_ns.ns);
+	if (err) {
+		Py_DECREF(iterator);
+		return set_drgn_error(err);
+	}
+	iterator->prog = prog_or_ns.prog;
+	Py_INCREF(iterator->prog);
```

Comment:
Missing a call to `prog_or_ns_cleanup` here, and also for all the error cases in this function.